### PR TITLE
MBID mapping

### PR DESCRIPTION
**Pull request by mayhem**
Due to the many changes that are happening right now, several features all ended up on the same branch leading to a longer PR than usual -- sorry for that.

In this PR:

- Drop the old MSID mapping concept
- Rename ACRP to MBID mapping and use typesense
- Implement typesense support and typesense based tests
- Due to a merge conflict recording_lookup needed to be merged/debugged again.
- Remove the buggy year support from the recording lookup endpoint.
- Remove mapping stats (no longer needed)
- Rename msb_mapping dir to mbid_mapping

### CODE REVIEW

File: listenbrainz/mbid_mapping/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+To create the MusicBrainz year mapping run:
+
+```manage.py year-mapping```
+
+
+Creating a typesense index for searching the mapping
+----------------------------------------------------
+
+If you plan to create a typesense index, you'll need typesense installed. First set an API KEY in config.py,
```
Comment (alastair) [inline listenbrainz/mbid_mapping/README.md:43]: can this api key be anything?

File: listenbrainz/mbid_mapping/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+then run:
+
+```
+docker run -p 8108:8108 -d -v <some_data_dir_for_typesense>/data --name=typesense
```
Comment (alastair) [inline listenbrainz/mbid_mapping/README.md:47]: there are now 2 different instructions here for running docker stuff independently. Can this be integrated either into a docker-compose file for mbid-mapping, or into the main LB docker-compose file?

File: listenbrainz/mbid_mapping/README.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+2. The same mapping as above, but focused on having the very first release of any given (artist_credit, recording_name)
+so that it can be used to find the first release year of that pair.
+
+3. A typesense index that takes the data from #1 and generates an index that can be used to quickly and fuzzily
+look up data from the table in #1. This index will serve as the official "MusicBrainz mapping".
+
+Creating mappings and running tests
+-----------------------------------
+
+You will need to have a copy of musicbrainz-docker installed and running in order to run these commands.
```
Comment (alastair) [inline listenbrainz/mbid_mapping/README.md:20]: link to musicbrainz-docker

File: listenbrainz/mbid_mapping/mapping/mapping_test/mapping_test.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+                else:
+                    continue
+
+            data.append(row)
+
+    return data
+
+
+def test_mapping():
+    ''' This test will run as many test as there are in the CSV file '''
```
Comment (alastair) [inline listenbrainz/mbid_mapping/mapping/mapping_test/mapping_test.py:24]: brief description of the format of the csv file

File: listenbrainz/mbid_mapping/mapping/mapping_test/mapping_test.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    passed = 0
+    failed = 0
+
+    for row in data:
+        hits = search(row[1] + " " + row[0])
+        if row[2] != hits[0]['release_mbid']:
+            print("Q %-30s %-30s %-25 %s" % (row[0][:29], "", row[1][:29], row[2]))
+            print("H %-30s %-30s %-25 %s" % (hits[0]['recording_name'][:29], hits[0]['release_name'][:29],
+                  hits[0]['artist_credit_name'][:29], hits[0]['release_mbid']))
+            print
```
Comment (alastair) [inline listenbrainz/mbid_mapping/mapping/mapping_test/mapping_test.py:37]: does this do anything?

File: listenbrainz/mbid_mapping/mapping/mapping_test/mapping_test.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            failed += 1
```
Comment (alastair) [inline listenbrainz/mbid_mapping/mapping/mapping_test/mapping_test.py:38]: do you want this to report the total number of passes and failures at the end of running, or could you use pytest here and just `assert` it? (It'll quit after the first failure)

File: listenbrainz/mbid_mapping/mapping/search.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+          'port': config.TYPESENSE_PORT,
+          'protocol': 'http',
+        }],
+        'api_key': config.TYPESENSE_API_KEY,
+        'connection_timeout_seconds': 2
+    })
+
+    search_parameters = {
+        'q': prepare_string(query),
+        'query_by': "combined",
```
Comment (alastair) [inline listenbrainz/mbid_mapping/mapping/search.py:23]: eventually it might be nice to move these parameters to config.py if we still want to be able to tweak them

File: listenbrainz/mbid_mapping/mapping/year_mapping.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
+import psycopg2
```
Comment (alastair) [inline listenbrainz/mbid_mapping/mapping/year_mapping.py:1]: Is this mostly the same as mbid_mapping.py, except with a slightly different initial query sort order? Can all of the common code be shared?

File: listenbrainz/mbid_mapping/README.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+
+To build the docker image for the mapping tools:
```
Comment (alastair) [inline listenbrainz/mbid_mapping/README.md:23]: copy config.py.sample to config.py

File: listenbrainz/mbid_mapping/push.sh [hunk 10]
```diff
@@ -10,5 +10,5 @@
```
Comment (alastair) [inline listenbrainz/mbid_mapping/push.sh:10]: Is this file still needed? (that is, do you use it to push the image to docker hub?) If so, please update the comment on line 4 (not listenbrainz) and line 5 (not 'latest')

**Review by alastair — CHANGES_REQUESTED** [review]
sorry for the delay, should be good now

File: listenbrainz/labs_api/labs/api/recording_from_recording_mbid.py [hunk 11]
```diff
@@ -105,19 +102,27 @@ def fetch(self, params, offset=-1, count=-1):
...
+                                           'length': None,
+                                           'comment': None,
+                                           'artist_credit_id': None,
+                                           'artist_credit_name': None,
+                                           '[artist_credit_mbids]': None,
+                                           'original_recording_mbid': mbid})
                             continue
 
-                    r['[artist_credit_mbids]'] = list(set(r['artist_credit_mbids']))
+                    r['[artist_credit_mbids]'] = [r for r in r['artist_credit_mbids']]
```
[the remark this answers is not in the document]
Comment (mayhem) [reply listenbrainz/labs_api/labs/api/recording_from_recording_mbid.py:120]: Amazing that it works in the first place!

File: listenbrainz/mbid_mapping/HACKING.md [hunk 12]
```diff
@@ -0,0 +1,23 @@
+                        SELECT substring(musicbrainz.musicbrainz_unaccent(ac.name) from 0 for 40) AS artist_credit_name,
```
[the remark this answers is not in the document]
Comment (mayhem) [reply listenbrainz/mbid_mapping/HACKING.md:1]: My commonly used commands scratchpad. I'll clean up.

File: listenbrainz/mbid_mapping/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,14 @@
+FROM ubuntu:20.10
```
[the remark this answers is not in the document]
Comment (mayhem) [reply listenbrainz/mbid_mapping/Dockerfile:1]: There may have been a reason, but I forget. I lumped all of this into "deployment tasks" (i.e. the next PR), but I'll update this now.

File: listenbrainz/mbid_mapping/README.md (hunk 1 above)
Comment (mayhem) [reply listenbrainz/mbid_mapping/README.md:43]: yep. kinda confusing.

File: listenbrainz/mbid_mapping/README.md (hunk 2 above)
Comment (mayhem) [reply listenbrainz/mbid_mapping/README.md:47]: I'll leave cleaning this up for the next PR.

File: listenbrainz/mbid_mapping/mapping/mapping_test/mapping_test.py (hunk 6 above)
Comment (mayhem) [reply listenbrainz/mbid_mapping/mapping/mapping_test/mapping_test.py:38]: I fail to see, (ha!) why it would quit after the first failure. Previously testing was more complicated and setting up pytest was a pita, but I can do this in a future PR. "good enough" for this moment.

File: listenbrainz/mbid_mapping/mapping/search.py (hunk 7 above)
Comment (mayhem) [reply listenbrainz/mbid_mapping/mapping/search.py:23]: :+1:

File: listenbrainz/mbid_mapping/mapping/year_mapping.py (hunk 8 above)
Comment (mayhem) [reply listenbrainz/mbid_mapping/mapping/year_mapping.py:1]: There are more subtle far-reaching changes than just that, but there is some code overlap that can be consolidated in a future PR.

File: listenbrainz/mbid_mapping/push.sh (hunk 10 above)
Comment (mayhem) [reply listenbrainz/mbid_mapping/push.sh:10]: Yep. Will update.

**Comment by mayhem** [comment]
I'm now fighting with jenkins test setup -- the tests pass locally and I've made fixes for this PR, so I am merging to be able to continue on finally.
